### PR TITLE
Bring the Swift frontend's handling of DIFiles in synch with CFE.

### PR DESCRIPTION
**Pull request by adrian-prantl**
This applies the same changes from Clang CFE r349065 to the Swift
frontend to unify how filenames, cmpilation directories and absolute
paths in filenames and path remappings are handled.

### CODE REVIEW

**Comment by adrian-prantl** [comment]
@swift-ci test and merge

**Comment by adrian-prantl** [comment]
@swift-ci test and merge

File: test/DebugInfo/basic.swift [hunk 1]
```diff
@@ -3,6 +3,7 @@
 // Verify that we don't emit any debug info by default.
 // RUN: %target-swift-frontend %s -emit-ir -o - \
 // RUN:   | %FileCheck %s --check-prefix NDEBUG
+// NDEBUG: foo
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply test/DebugInfo/basic.swift]: The FileCheck interpretation of
```
CHECK: a
CHECK-NOT: b
CHECK-NOT: c
```
is "find 'a', then make sure that I don't find 'b' or 'c' after it". It doesn't stop 'b' and 'c' from appearing *before* 'a'.

[the remark this answers is not in the document]
Comment (adrian-prantl) [reply test/DebugInfo/basic.swift]: Thanks for clarifying! That's actually what I intended effect here: I just wanted to add a positive check to ensure that we emit LLVM IR at all, in this case the function definition and then make sure no !dbg attachment appears in the function body after that. Since `foo` is a terrible string to search for I changed it to source_filename now, which is guaranteed to appear at the very top of the output.

[the remark this answers is not in the document]
Comment (jrose-apple) [reply test/DebugInfo/basic.swift]: Ah, okay! Sorry, seen too many cases where I add a CHECK line and miss that I just allowed a negative pattern to slip in before it.

**Comment by adrian-prantl** [comment]
@swift-ci test and merge

**Comment by adrian-prantl** [comment]
(Fixed and edge case that required the SDK path to have zero common components with the compilation directory, such as Xcode installed in /Application to reproduce)

**Comment by adrian-prantl** [comment]
@swift-ci test
